### DISCUSSION TITLE
`vi.mock` not working as expected.

### DISCUSSION THREAD

**Discussion by robbporto** [vitest-dev/vitest#5334] [Q&A] [answered]
### The component

As an example, I have the following jest:

```js
import { render, waitFor } from '@testing-library/react'

vi.mock('react-router-dom')

it('should do something', () => {
  const { getByText } = render(<Component />)

  expect(getByText('Something')).toBeInTheDocument()
})
```

`<Component />` is using `react-router-dom` so I need to mock it, but this test is not working (all functions from `react-router-dom` are returning `undefined`). My idea was that creating a file inside the `src` folder with the same name of the node module would work, but it is not.

### The mock

Here's how my mock looks like:

```js
// src/__mocks__/react-router-dom.js

import { vi } from 'vitest'

const mockUseNavigate = vi.fn()

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom')
  return {
    ...actual,
    useNavigate: () => mockUseNavigate,
    Link: vi.fn(),
    useParams: () => ({ id: 123 })
  }
})
export { mockUseNavigate }

```

### Alternative 1 (not optimal)

If I mock `react-router-dom` inside the test instead, it works. Like this:

```js
import { render, waitFor } from '@testing-library/react'
import {vi} from 'vitest'

const mockUseNavigate = vi.fn()

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom')
  return {
    ...actual,
    useNavigate: () => mockUseNavigate,
    Link: vi.fn(),
    useParams: () => ({ id: 123 })
  }
})


it('should do something', () => {
  const { getByText } = render(<Component />)

  expect(getByText('Something')).toBeInTheDocument()
})
```

### Alternative 2 (not optimal)

Also, if I import the mock instead of using `vi.mock`, it also works:

```js
import * as reactRouterDom from '../../__mocks__/react-router-dom.js'
import { render, waitFor } from '@testing-library/react'
...
```

## Question
Alternatives 1 and 2 are not optimal. I don't want to have to mock `react-router-dom` in every file and sometimes I don't need to use any function from `reactRouterDom`, so it doesn't make sense importing it.

I would like to just use `vi.mock('react-router-dom')`, but clearly I'm doing something wrong, but I don't know what. I would appreciate some help on this! Thank you!

**Comment by sheremet-va** [comment]
Everything you described works as expected: https://vitest.dev/guide/mocking.html#automocking-algorithm

**Reply by robbporto** [reply]
> If your code is importing a mocked module, without any associated __mocks__ file or factory for this module, Vitest will mock the module itself by invoking it and mocking every export.

Thank you for answering.

I do have an associated __mocks__ file (src/__mocks__/react-router-dom.js) with the same name of the library that I want mocked (react-router-dom).

Can you elaborate what you mean, please?

**Reply by sheremet-va** [reply]
1. Is `src` set as a [root](https://vitest.dev/config/#root)?
2. Why does it call vi.mock?

**Reply by robbporto** [reply]
1. `src` is not set as root in the `vite.config.js` file.
   - if I set `src` as root and move the `__mocks__` folder to the folder above `src` (the real root folder), the tests pass;
   - if I set `src` as root and move the `__mocks__` folder to the folder above `src` (the real root folder), the tests fail;
   
   I'm not sure what's going on here.

2. Are you talking about the mock file? I thought that's how you're supposed to mock files with `vitest`. I mock external libraries (node modules) using `jest.mock` for Jest, I thought it would be the same for Vitest.

**Reply by sheremet-va** [reply, marked as answer]
Files under `__mocks__` are picked up automatically as long as you call `vi.mock(libraryName)` _outside_ of that file - otherwise mocker doesn't even know it should be mocked. It should reexport the same named export that you have in your library. If you need to have it mocked in every file, use `setupFiles` (https://vitest.dev/api/vi.html#vi-mock):

> If you are mocking a dependency, Vitest will try to find a `__mocks__` folder in the [root](https://vitest.dev/config/#root) of the project (default is process.cwd()). You can tell Vitest where the dependencies are located through the [deps.moduleDirectories](https://vitest.dev/config/#deps-moduledirectories) config option.
> If you call vi.mock in a test file without a factory provided, it will find a file in the __mocks__ folder to use as a module
> Beware that if you don't call vi.mock, modules **are not** mocked automatically. To replicate Jest's automocking behaviour, you can call vi.mock for each required module inside [setupFiles](https://vitest.dev/config/#setupfiles).

Please, refer to documentation to understand how mocking works.

**Reply by robbporto** [reply]
Hi, @sheremet-va. Thanks for the answers. I wasn't able to mock the library the way I wanted, but what I did for now was to move the mock to `setupFiles` like you suggested and that's OK for now.

Thank you for the help.
